### ISSUE TITLE
deadlock occurs when create multiple indexes on AO table at the same time

### ISSUE THREAD

**Issue by LittleWuCoding** [greenplum-db/gpdb#10961]
### Greenplum version or build
GP6.10 （I did not test the master branch）
### OS version and uname -a
RHEL7

### Step to reproduce the behavior
1. create a AO table , for example (same error occurs on AOCO ):
```
CREATE table t1(c1 int,c2 text, c3 text, c4 int)
with (appendonly=true,orientation=row,
compresstype=ZLIB, compresslevel=5)
distributed by (c1);
```
2. create multiple indexes on AO table at the same time:
```
# statement1
psql -d postgres -c "create index t1_xx21 ON t1(c2);" &
# statement2
psql -d postgres -c "create index t1_xx22 ON t1(c1,c2);" &
# statement3
psql -d postgres -c "create index t1_xx23 ON t1(c2,c1);" &
```
(you may also reproduce by only exec statement1 and statement3, but not always )
I put the above cmd in a shell script, named test1.sh run it:
```
sh test1.sh
```
Output the following error:
```
CREATE INDEX
ERROR:  deadlock detected
DETAIL:  Process 175290 waits for ShareRowExclusiveLock on relation 16384 of database 12806; blocked by process 175291.
Process 175291 waits for ShareRowExclusiveLock on relation 16384 of database 12806; blocked by process 175290.
HINT:  See server log for query details.
CREATE INDEX
```
IN the error message, relation 16384 is the AO table t1
I also test the heap table, which Successfully executed without error.
So, I think it is a bug related to AO table, I have a glance on the source, found that in function DefineIndex() defined in file src/backend/commands/indexcmds.c, the lock aquired by AO table is different from heap table:
```
if (RangeVarIsAppendOptimizedTable(stmt->relation))
		lockmode = ShareRowExclusiveLock;
	else
		lockmode = stmt->concurrent ? ShareUpdateExclusiveLock : ShareLock;
```
Also, I gdb the program(just for reference， maybe not the root cause),get the below stacks:
```
#0  0x00007fc52bac40d7 in semop () from /lib64/libc.so.6
#1  0x00000000008c704b in PGSemaphoreLock (sema=0x7fc51fc82710, interruptOK=1 '\001') at pg_sema.c:422
#2  0x000000000095834e in ProcSleep (locallock=0x2e14780, lockMethodTable=0xdcfc20 <default_lockmethod>) at proc.c:1347
#3  0x0000000000950f7b in WaitOnLock (locallock=0x2e14780, owner=0x2e66878) at lock.c:1853
#4  0x000000000094fcfa in LockAcquireExtended (locktag=0x7ffff2075750, lockmode=5, sessionLock=0 '\000', dontWait=0 '\000', reportMemoryError=1 '\001', 
    locallockp=0x7ffff2075748) at lock.c:1155
#5  0x000000000094ce5d in LockRelationOid (relid=16384, lockmode=5) at lmgr.c:105
#6  0x00000000005c244e in RangeVarGetRelidExtended (relation=0x2e07c18, lockmode=5, missing_ok=0 '\000', nowait=0 '\000', 
    callback=0x6fc8da <RangeVarCallbackOwnsRelation>, callback_arg=0x0) at namespace.c:390
#7  0x00000000009792cc in ProcessUtilitySlow (parsetree=0x2e07de8, queryString=0x2e07128 "create index xx22 ON t1(c1,c2);", context=PROCESS_UTILITY_TOPLEVEL, 
    params=0x0, dest=0x2e081d0, completionTag=0x7ffff2075d70 "") at utility.c:1511
#8  0x00000000009785c4 in standard_ProcessUtility (parsetree=0x2e07de8, queryString=0x2e07128 "create index xx22 ON t1(c1,c2);", context=PROCESS_UTILITY_TOPLEVEL, 
    params=0x0, dest=0x2e081d0, completionTag=0x7ffff2075d70 "") at utility.c:1039
#9  0x00000000009774f4 in ProcessUtility (parsetree=0x2e07de8, queryString=0x2e07128 "create index xx22 ON t1(c1,c2);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, 
    dest=0x2e081d0, completionTag=0x7ffff2075d70 "") at utility.c:374
#10 0x00000000009762c7 in PortalRunUtility (portal=0x2e0ee08, utilityStmt=0x2e07de8, isTopLevel=1 '\001', dest=0x2e081d0, completionTag=0x7ffff2075d70 "")
    at pquery.c:1381
#11 0x0000000000976558 in PortalRunMulti (portal=0x2e0ee08, isTopLevel=1 '\001', dest=0x2e081d0, altdest=0x2e081d0, completionTag=0x7ffff2075d70 "") at pquery.c:1512
#12 0x0000000000975a4f in PortalRun (portal=0x2e0ee08, count=9223372036854775807, isTopLevel=1 '\001', dest=0x2e081d0, altdest=0x2e081d0, 
    completionTag=0x7ffff2075d70 "") at pquery.c:1018
#13 0x000000000096df53 in exec_simple_query (query_string=0x2e07128 "create index xx22 ON t1(c1,c2);") at postgres.c:1824
#14 0x0000000000972d29 in PostgresMain (argc=1, argv=0x2de6d50, dbname=0x2de6bb0 "postgres", username=0x2de6b90 "whydb") at postgres.c:5246
#15 0x00000000008e008e in BackendRun (port=0x2e0e1a0) at postmaster.c:4811
#16 0x00000000008df72e in BackendStartup (port=0x2e0e1a0) at postmaster.c:4468
#17 0x00000000008db6cb in ServerLoop () at postmaster.c:1948
#18 0x00000000008dac1d in PostmasterMain (argc=6, argv=0x2de4c10) at postmaster.c:1518
#19 0x00000000007d406c in main (argc=6, argv=0x2de4c10) at main.c:245
```
Hope the above information can help~

**Comment by kainwen** [comment]
This deadlock on QD is due to lock mode upgrade. We should avoid this.

**Comment by LittleWuCoding** [comment]
Hi, Is there any news about this issue? @kainwen @gaos1

**Comment by kainwen** [comment]
> Hi, Is there any news about this issue? @kainwen @gaos1

There is fixme at the code for this issue. We will resolve the fixme later.

**Comment by ashwinstar** [comment]
I think to resolve this issue we need to create block directory upfront for AO tables. Currently, one first index creation block directory is created and that I suppose is causing this lock upgrade issue. One of the solutions discuss in past was create block directory when creating AO tables instead of differing creation but impact is tables not using index its waste plus also poses upgrade time work. I don't know what other way it can be resolved.

**Comment by d** [comment]
> I think to resolve this issue we need to create block directory upfront for AO tables.

Yes I've been lobbying for this for the last 2 years. Keep in mind we'll need to be careful with upgrade if we pursue that

**Comment by ashwinstar** [comment]
Just to mention one other data point, for CO tables just due to deferred block directory creation first index creation needs to scan all the columns (instead of just indexed columns) to create block directory. If block directory already exists on required indexed columns are scanned which makes it much faster.

**Comment by ashwinstar** [comment]
We have no plan to fix this issue as requires significant design decision change to create block directory upfront. Reward from fix vs cost of fixing the issue, currently hints no fix as not significant pain point. Various workarounds exist for the problem like creating the index and dropping the index just after creating a table where wish to create multiple indexes at same time later on the table. This would create the block-directory upfront for the table.
